Sources/MessagePack/MessagePackEncoder.swift

```diff
@@ -0,0 +1,1514 @@
+        return try unbox(storage.topContainer, as: Float.self)!
+    }
+    
+    public func decode(_ type: Double.Type) throws -> Double {
+        return try unbox(storage.topContainer, as: Double.self)!
+    }
+    
+    public func decode(_ type: String.Type) throws -> String {
+        return try unbox(storage.topContainer, as: String.self)!
+    }
+    
+    public func decode<T : Decodable>(_ type: T.Type) throws -> T {
+        return try unbox(storage.topContainer, as: T.self)!
+    }
+}
+
+// MARK: Concrete Value Representations
+
+extension _MessagePackDecoder {
+    
+    fileprivate func unbox(_ value: MessagePackValue, as type: Bool.Type) throws -> Bool? {
+        return value.boolValue
+    }
+    
+    fileprivate func unbox(_ value: MessagePackValue, as type: Int.Type) throws -> Int? {
+        switch options.numberDecodingStrategy {
+        case .noTypeConversion:         return value.integerValue.map { Int($0) }
+        case .automaticTypeConversion:  return value.numberValue?.intValue
+        }
+    }
+    
+    fileprivate func unbox(_ value: MessagePackValue, as type: Int8.Type) throws -> Int8? {
+        switch options.numberDecodingStrategy {
+        case .noTypeConversion:         return value.integerValue.map { Int8($0) }
+        case .automaticTypeConversion:  return value.numberValue?.int8Value
+        }
+    }
+    
+    fileprivate func unbox(_ value: MessagePackValue, as type: Int16.Type) throws -> Int16? {
+        switch options.numberDecodingStrategy {
+        case .noTypeConversion:         return value.integerValue.map { Int16($0) }
+        case .automaticTypeConversion:  return value.numberValue?.int16Value
+        }
+    }
+    
+    fileprivate func unbox(_ value: MessagePackValue, as type: Int32.Type) throws -> Int32? {
+        switch options.numberDecodingStrategy {
+        case .noTypeConversion:         return value.integerValue.map { Int32($0) }
+        case .automaticTypeConversion:  return value.numberValue?.int32Value
+        }
+    }
+    
+    fileprivate func unbox(_ value: MessagePackValue, as type: Int64.Type) throws -> Int64? {
+        switch options.numberDecodingStrategy {
+        case .noTypeConversion:         return value.integerValue.map { Int64($0) }
+        case .automaticTypeConversion:  return value.numberValue?.int64Value
+        }
+    }
+    
+    fileprivate func unbox(_ value: MessagePackValue, as type: UInt.Type) throws -> UInt? {
+        switch options.numberDecodingStrategy {
+        case .noTypeConversion:         return value.unsignedIntegerValue.map { UInt($0) }
+        case .automaticTypeConversion:  return value.numberValue?.uintValue
+        }
+    }
+    
+    fileprivate func unbox(_ value: MessagePackValue, as type: UInt8.Type) throws -> UInt8? {
+        switch options.numberDecodingStrategy {
+        case .noTypeConversion:         return value.unsignedIntegerValue.map { UInt8($0) }
```

Comment:
This type cast will cause an EXC_BAD_INSTRUCTION if the underlying integer is larger than UInt8.max. The same applies for every other implementation of `unbox` that casts to integers smaller than Int64 (including Int which should not be assumed to be 64-bits).

Reply:
Right, i'll fix this and add test cases for this